src/rgw/rgw_s3select.cc

```diff
@@ -655,6 +714,70 @@ int RGWSelectObj_ObjStore_S3::csv_processing(bufferlist& bl, off_t ofs, off_t le
   return status;
 }
 
+int RGWSelectObj_ObjStore_S3::json_processing(bufferlist& bl, off_t ofs, off_t len)
+{
+  int status = 0;
+
+  const char* s3select_syntax_error = "s3select-Syntax-Error";
+  const char* s3select_resource_id = "resourcse-id";
+  const char* s3select_json_error = "json-Format-Error";
+
+  s3select_syntax.parse_query(m_sql_query.c_str());
+  m_s3_json_object.set_json_query(&s3select_syntax);
+
+  if (output_row_delimiter.size()) {
+    output_row_delimiter = *output_row_delimiter.c_str();
+  }
+  
+  if(m_json_datatype.compare("DOCUMENT") != 0) {
+    m_aws_response_handler.send_error_response(s3select_json_error,
+        s3select_syntax.get_error_description().c_str(),
+        s3select_resource_id);
+    ldpp_dout(this, 10) << "s3-select query: wrong json format; {" << s3select_syntax.get_error_description() << "}" << dendl;
+    return -1;
+  } 
+
+  if (s3select_syntax.get_error_description().empty() == false) {
+    //error-flow (syntax-error)
+    m_aws_response_handler.send_error_response(s3select_syntax_error,
+        s3select_syntax.get_error_description().c_str(),
+        s3select_resource_id);
+    ldpp_dout(this, 10) << "s3-select query: failed to prase query; {" << s3select_syntax.get_error_description() << "}" << dendl;
+    return -1;
+  }
+  
+  if (s->obj_size == 0) {
+    status = run_s3select_on_json(m_sql_query.c_str(), nullptr, 0);
+  } else {
+    auto bl_len = bl.get_num_buffers();
```

Comment:
`status` is unchecked here, then we immediately go on to call `status = run_s3select_on_json(...)` below

do we need this special case for `obj_size == 0` at all? it should be fine to loop over an empty `bl.buffers()`

Reply:
as for the `obj_size == 0` i recall we needed a specific flow for CSV upon zero size object.
it will be re-check.

Reply:
> `status` is unchecked here, then we immediately go on to call `status = run_s3select_on_json(...)` below
> 
> do we need this special case for `obj_size == 0` at all? it should be fine to loop over an empty `bl.buffers()`

upon empty-object(zero size), the `bl = buffer-list` is empty, thus buffer-list loop does not execute, no call to s3select-function, which results with a wrong response.

`if (s->obj_size == 0) { ... } `  returns the appropriate empty response.